File: scripts/podcaster-prototype.js
#!/usr/bin/env node

/**
 * Podcaster Prototype - Issue #214
 * Converts markdown files to audio using edge-tts
 */

import { ttsSave } from 'edge-tts';
import { readFile } from 'fs/promises';
import { stat } from 'fs/promises';
import { resolve, basename, extname } from 'path';

// Strip markdown formatting to plain text
function stripMarkdown(markdown) {
  let text = markdown;
  
  // Remove YAML frontmatter
  text = text.replace(/^---\n[\s\S]*?\n---\n/m, '');
  
  // Remove HTML comments
  text = text.replace(/<!--[\s\S]*?-->/g, '');
  
  // Remove code blocks
  text = text.replace(/```[\s\S]*?```/g, '');
  text = text.replace(/`[^`]+`/g, '');
  
  // Remove images
  text = text.replace(/!\[([^\]]*)\]\([^)]+\)/g, '$1');
  
  // Remove links but keep text
  text = text.replace(/\[([^\]]+)\]\([^)]+\)/g, '$1');
  
  // Remove headers but keep text
  text = text.replace(/^#{1,6}\s+/gm, '');
  
  // Remove bold/italic
  text = text.replace(/\*\*([^*]+)\*\*/g, '$1');
  text = text.replace(/\*([^*]+)\*/g, '$1');
  text = text.replace(/__([^_]+)__/g, '$1');
  text = text.replace(/_([^_]+)_/g, '$1');
  
  // Remove horizontal rules
  text = text.replace(/^[-*_]{3,}\s*$/gm, '');
  
  // Remove blockquotes
  text = text.replace(/^>\s+/gm, '');
  
  // Remove list markers
  text = text.replace(/^[\*\-\+]\s+/gm, '');
  text = text.replace(/^\d+\.\s+/gm, '');
  
  // Clean up multiple newlines
  text = text.replace(/\n{3,}/g, '\n\n');
  
  // Trim whitespace
  text = text.trim();
  
  return text;
}

// Format file size
function formatBytes(bytes) {
  if (bytes < 1024) return bytes + ' B';
  if (bytes < 1024 * 1024) return (bytes / 1024).toFixed(2) + ' KB';
  return (bytes / (1024 * 1024)).toFixed(2) + ' MB';
}

// Estimate audio duration (rough approximation: ~150 words per minute)
function estimateDuration(text) {
  const words = text.split(/\s+/).length;
  const minutes = words / 150;
  const seconds = Math.round(minutes * 60);
  
  if (seconds < 60) return `${seconds}s`;
  const mins = Math.floor(seconds / 60);
  const secs = seconds % 60;
  return `${mins}m ${secs}s`;
}

async function main() {
  const args = process.argv.slice(2);
  
  if (args.length === 0) {
    console.error('Usage: node podcaster-prototype.js <markdown-file>');
    console.error('Example: node podcaster-prototype.js RESEARCH_REPORT.md');
    process.exit(1);
  }
  
  const inputPath = resolve(args[0]);
  const inputFilename = basename(inputPath, extname(inputPath));
  const outputPath = resolve(`${inputFilename}-audio.mp3`);
  
  console.log('🎙️  Podcaster Prototype - Issue #214\n');
  console.log(`📄 Input: ${inputPath}`);
  console.log(`🔊 Output: ${outputPath}\n`);
  
  try {
    // Read markdown file
    console.log('📖 Reading markdown file...');
    const markdown = await readFile(inputPath, 'utf-8');
    console.log(`   Markdown size: ${formatBytes(markdown.length)}`);
    
    // Strip markdown formatting
    console.log('🔧 Stripping markdown formatting...');
    const plainText = stripMarkdown(markdown);
    console.log(`   Plain text size: ${formatBytes(plainText.length)}`);
    console.log(`   Estimated duration: ${estimateDuration(plainText)}`);
    
    // Convert to speech
    console.log('🎤 Converting to speech (using edge-tts)...');
    const startTime = Date.now();
    
    await ttsSave(plainText, outputPath, {
      voice: 'en-US-JennyNeural',  // Professional female voice
      rate: '+0%',
      pitch: '+0Hz'
    });
    
    const endTime = Date.now();
    const conversionTime = ((endTime - startTime) / 1000).toFixed(2);
    
    // Get output file stats
    const stats = await stat(outputPath);
    
    console.log(`✅ Conversion complete in ${conversionTime}s`);
    console.log(`\n📊 Results:`);
    console.log(`   Audio file: ${outputPath}`);
    console.log(`   File size: ${formatBytes(stats.size)}`);
    console.log(`   Format: MP3`);
    console.log(`   Voice: en-US-JennyNeural (Microsoft Neural TTS)`);
    console.log(`   Quality: Neural (production-grade)`);
    console.log(`\n✨ Success! Audio file generated.`);
    
  } catch (error) {
    console.error(`\n❌ Error: ${error.message}`);
    process.exit(1);
  }
}

main();

File: scripts/podcaster-prototype.py
#!/usr/bin/env python3

"""
Podcaster Prototype - Issue #214
Converts markdown files to audio using edge-tts
"""

import asyncio
import re
import sys
import os
from pathlib import Path
import edge_tts

def strip_markdown(markdown):
    """Strip markdown formatting to plain text"""
    text = markdown
    
    # Remove YAML frontmatter
    text = re.sub(r'^---\n[\s\S]*?\n---\n', '', text, flags=re.MULTILINE)
    
    # Remove HTML comments
    text = re.sub(r'<!--[\s\S]*?-->', '', text)
    
    # Remove code blocks
    text = re.sub(r'```[\s\S]*?```', '', text)
    text = re.sub(r'`[^`]+`', '', text)
    
    # Remove images but keep alt text
    text = re.sub(r'!\[([^\]]*)\]\([^)]+\)', r'\1', text)
    
    # Remove links but keep text
    text = re.sub(r'\[([^\]]+)\]\([^)]+\)', r'\1', text)
    
    # Remove headers but keep text
    text = re.sub(r'^#{1,6}\s+', '', text, flags=re.MULTILINE)
    
    # Remove bold/italic
    text = re.sub(r'\*\*([^*]+)\*\*', r'\1', text)
    text = re.sub(r'\*([^*]+)\*', r'\1', text)
    text = re.sub(r'__([^_]+)__', r'\1', text)
    text = re.sub(r'_([^_]+)_', r'\1', text)
    
    # Remove horizontal rules
    text = re.sub(r'^[-*_]{3,}\s*$', '', text, flags=re.MULTILINE)
    
    # Remove blockquotes
    text = re.sub(r'^>\s+', '', text, flags=re.MULTILINE)
    
    # Remove list markers
    text = re.sub(r'^[\*\-\+]\s+', '', text, flags=re.MULTILINE)
    text = re.sub(r'^\d+\.\s+', '', text, flags=re.MULTILINE)
    
    # Clean up multiple newlines
    text = re.sub(r'\n{3,}', '\n\n', text)
    
    # Trim whitespace
    text = text.strip()
    
    return text

def format_bytes(bytes_size):
    """Format file size"""
    if bytes_size < 1024:
        return f"{bytes_size} B"
    elif bytes_size < 1024 * 1024:
        return f"{bytes_size / 1024:.2f} KB"
    else:
        return f"{bytes_size / (1024 * 1024):.2f} MB"

def estimate_duration(text):
    """Estimate audio duration (rough: ~150 words per minute)"""
    words = len(text.split())
    minutes = words / 150
    seconds = round(minutes * 60)
    
    if seconds < 60:
        return f"{seconds}s"
    mins = seconds // 60
    secs = seconds % 60
    return f"{mins}m {secs}s"

async def main():
    if len(sys.argv) < 2:
        print("Usage: python podcaster-prototype.py <markdown-file>")
        print("Example: python podcaster-prototype.py RESEARCH_REPORT.md")
        sys.exit(1)
    
    input_path = Path(sys.argv[1]).resolve()
    input_filename = input_path.stem
    output_path = Path(f"{input_filename}-audio.mp3").resolve()
    
    print("🎙️  Podcaster Prototype - Issue #214\n")
    print(f"📄 Input: {input_path}")
    print(f"🔊 Output: {output_path}\n")
    
    try:
        # Read markdown file
        print("📖 Reading markdown file...")
        with open(input_path, 'r', encoding='utf-8') as f:
            markdown = f.read()
        print(f"   Markdown size: {format_bytes(len(markdown.encode('utf-8')))}")
        
        # Strip markdown formatting
        print("🔧 Stripping markdown formatting...")
        plain_text = strip_markdown(markdown)
        print(f"   Plain text size: {format_bytes(len(plain_text.encode('utf-8')))}")
        print(f"   Estimated duration: {estimate_duration(plain_text)}")
        
        # Convert to speech
        print("🎤 Converting to speech (using edge-tts)...")
        import time
        start_time = time.time()
        
        # Use Jenny Neural voice (professional female voice)
        communicate = edge_tts.Communicate(plain_text, "en-US-JennyNeural")
        await communicate.save(str(output_path))
        
        end_time = time.time()
        conversion_time = end_time - start_time
        
        # Get output file stats
        file_size = output_path.stat().st_size
        
        print(f"✅ Conversion complete in {conversion_time:.2f}s")
        print(f"\n📊 Results:")
        print(f"   Audio file: {output_path}")
        print(f"   File size: {format_bytes(file_size)}")
        print(f"   Format: MP3")
        print(f"   Voice: en-US-JennyNeural (Microsoft Neural TTS)")
        print(f"   Quality: Neural (production-grade)")
        print(f"\n✨ Success! Audio file generated.")
        
    except FileNotFoundError:
        print(f"\n❌ Error: File not found: {input_path}")
        sys.exit(1)
    except Exception as error:
        print(f"\n❌ Error: {error}")
        sys.exit(1)

if __name__ == "__main__":
    asyncio.run(main())
